pkg/agent/openflow/multicluster.go

```diff
@@ -0,0 +1,233 @@
+)
+
+// GlobalVirtualMACForMulticluster is a vritual MAC which will be used only
+// for cross-cluster traffic to distinguish from in-cluster traffic.
+var GlobalVirtualMACForMulticluster, _ = net.ParseMAC("aa:bb:cc:dd:ee:f0")
+
+type featureMulticluster struct {
+	cookieAllocator cookie.Allocator
+	cachedFlows     *flowCategoryCache
+	category        cookie.Category
+	ipProtocols     []binding.Protocol
+	dnatCtZones     map[binding.Protocol]int
+	snatCtZones     map[binding.Protocol]int
+}
+
+func (f *featureMulticluster) getFeatureName() string {
+	return "Multicluster"
+}
+
+func newFeatureMulticluster(cookieAllocator cookie.Allocator, ipProtocols []binding.Protocol) *featureMulticluster {
+	snatCtZones := make(map[binding.Protocol]int)
+	dnatCtZones := make(map[binding.Protocol]int)
+	snatCtZones[ipProtocols[0]] = SNATCtZone
+	dnatCtZones[ipProtocols[0]] = CtZone
+	return &featureMulticluster{
+		cookieAllocator: cookieAllocator,
+		cachedFlows:     newFlowCategoryCache(),
+		category:        cookie.Multicluster,
+		ipProtocols:     ipProtocols,
+		snatCtZones:     snatCtZones,
+		dnatCtZones:     dnatCtZones,
+	}
+}
+
+func (f *featureMulticluster) initFlows() []binding.Flow {
+	return []binding.Flow{}
+}
+
+func (f *featureMulticluster) replayFlows() []binding.Flow {
+	return getCachedFlows(f.cachedFlows)
+}
+
+func (f *featureMulticluster) l3FwdFlowToGatewayNodeViaTun(
+	localGatewayMAC net.HardwareAddr,
+	peerServiceCIDR net.IPNet,
+	tunnelPeer net.IP,
+	remoteGatewayIP net.IP) []binding.Flow {
+	ipProtocol := getIPProtocol(peerServiceCIDR.IP)
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+	flows = append(flows,
+		// This generates the flow to forward cross-cluster request packets based
+		// on Service ClusterIP range.
+		L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchProtocol(ipProtocol).
+			MatchDstIPNet(peerServiceCIDR).
+			Action().SetSrcMAC(localGatewayMAC).  // Rewrite src MAC to local gateway MAC.
+			Action().SetDstMAC(GlobalVirtualMAC). // Rewrite dst MAC to virtual MAC.
+			Action().SetTunnelDst(tunnelPeer).    // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done(),
+		// This generates the flow to forward cross-cluster reply traffic based
+		// on Gateway IP.
+		L3ForwardingTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
+			MatchProtocol(ipProtocol).
+			MatchCTStateRpl(true).
+			MatchCTStateTrk(true).
+			MatchDstIP(remoteGatewayIP).
+			Action().SetSrcMAC(localGatewayMAC).
+			Action().SetDstMAC(GlobalVirtualMACForMulticluster).
+			Action().SetTunnelDst(tunnelPeer). // Flow based tunnel. Set tunnel destination.
+			Action().LoadRegMark(ToTunnelRegMark).
+			Action().GotoTable(L3DecTTLTable.GetID()).
+			Done(),
+	)
+	return flows
+}
+
+func (f *featureMulticluster) l3FwdFlowToRemoteGatewayViaTun(
```

Comment:
Is this flow exactly same as above one except they use different tunnel IP which is already an argument? if yes, it seems they could be merged into one method like `l3FwdFlowToRemoteViaTun`, in one case the tunnel endpoint is local gateway, in another case the tunnel endpoint is remote gateway?

Reply:
A destination mac is different as well in one flow, but yes, they can be merged I have refined into one method.